From 0574339829f36e302f4bc05d49830fa73ae2c4f5 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <perex@perex.cz>
Date: Thu, 23 Jul 2020 12:43:48 +0800
Subject: [PATCH 11/14] HDA-Intel: add support for AMD acp microphone devices

- move the generic HDA code from sof-hda-dsp to HDA-Intel
- add generic codecs/hda/hdmi.conf

Signed-off-by: Jaroslav Kysela <perex@perex.cz>
(backported from commit dcef48f13d4f5db79b006755074940b94730a883)
Signed-off-by: Hui Wang <hui.wang@canonical.com>
---
 ucm2/HDA-Intel/HDA-Capture-value.conf |   4 +
 ucm2/HDA-Intel/HDA-Intel.conf         |  28 ++++++
 ucm2/HDA-Intel/Hdmi.conf              |  37 +++++++
 ucm2/HDA-Intel/HiFi-acp.conf          |   8 ++
 ucm2/HDA-Intel/HiFi-analog.conf       | 133 ++++++++++++++++++++++++++
 ucm2/HDA-Intel/HiFi.conf              |  32 +++++++
 ucm2/HDA-Intel/init.conf              |  39 ++++++++
 ucm2/codecs/hda/hdmi.conf             |  25 +++++
 ucm2/module/lib/linked.conf           |   3 +
 ucm2/module/snd_acp3x_rn.conf         |   1 +
 10 files changed, 310 insertions(+)
 create mode 100644 ucm2/HDA-Intel/HDA-Capture-value.conf
 create mode 100644 ucm2/HDA-Intel/HDA-Intel.conf
 create mode 100644 ucm2/HDA-Intel/Hdmi.conf
 create mode 100644 ucm2/HDA-Intel/HiFi-acp.conf
 create mode 100644 ucm2/HDA-Intel/HiFi-analog.conf
 create mode 100644 ucm2/HDA-Intel/HiFi.conf
 create mode 100644 ucm2/HDA-Intel/init.conf
 create mode 100644 ucm2/codecs/hda/hdmi.conf
 create mode 100644 ucm2/module/lib/linked.conf
 create mode 120000 ucm2/module/snd_acp3x_rn.conf

diff --git a/ucm2/HDA-Intel/HDA-Capture-value.conf b/ucm2/HDA-Intel/HDA-Capture-value.conf
new file mode 100644
index 0000000..2f1316a
--- /dev/null
+++ b/ucm2/HDA-Intel/HDA-Capture-value.conf
@@ -0,0 +1,4 @@
+CapturePCM "hw:${CardId}"
+CaptureMixerElem "Capture"
+CaptureVolume "Capture Volume"
+CaptureSwitch "Capture Switch"
diff --git a/ucm2/HDA-Intel/HDA-Intel.conf b/ucm2/HDA-Intel/HDA-Intel.conf
new file mode 100644
index 0000000..109eea6
--- /dev/null
+++ b/ucm2/HDA-Intel/HDA-Intel.conf
@@ -0,0 +1,28 @@
+Syntax 3
+
+Define.Use ""	# a non-empty string to use UCM configuration for HDA devices
+
+Define.AcpCardId "$${CardIdByName:acp}"
+
+If.acp {
+	Condition {
+		Type String
+		Empty "${var:AcpCardId}"
+	}
+	False.Define.Use y
+}
+
+If.use {
+	Condition {
+		Type String
+		Empty "${var:Use}"
+	}
+	False {
+		SectionUseCase."HiFi" {
+			File "HiFi.conf"
+			Comment "Play HiFi quality Music"
+		}
+
+		Include.init.File "/HDA-Intel/init.conf"
+	}
+}
diff --git a/ucm2/HDA-Intel/Hdmi.conf b/ucm2/HDA-Intel/Hdmi.conf
new file mode 100644
index 0000000..8b1843b
--- /dev/null
+++ b/ucm2/HDA-Intel/Hdmi.conf
@@ -0,0 +1,37 @@
+# Use case Configuration for sof-hda-dsp
+
+If.hdmi1 {
+	Condition { Type String Empty "" }
+	True {
+		Define {
+			HdmiNum 1
+			HdmiPCM 3
+			HdmiPrio 500
+		}
+		Include.hdmi1.File "/codecs/hda/hdmi.conf"
+	}
+}
+
+If.hdmi2 {
+	Condition { Type String Empty "" }
+	True {
+		Define {
+			HdmiNum 2
+			HdmiPCM 7
+			HdmiPrio 600
+		}
+		Include.hdmi2.File "/codecs/hda/hdmi.conf"
+	}
+}
+
+If.hdmi3 {
+	Condition { Type String Empty "" }
+	True {
+		Define {
+			HdmiNum 3
+			HdmiPCM 8
+			HdmiPrio 700
+		}
+		Include.hdmi3.File "/codecs/hda/hdmi.conf"
+	}
+}
diff --git a/ucm2/HDA-Intel/HiFi-acp.conf b/ucm2/HDA-Intel/HiFi-acp.conf
new file mode 100644
index 0000000..123ae15
--- /dev/null
+++ b/ucm2/HDA-Intel/HiFi-acp.conf
@@ -0,0 +1,8 @@
+SectionDevice."Mic1" {
+	Comment "Digital Microphone"
+
+	Value {
+		CapturePriority 100
+		CapturePCM "hw:${var:AcpCardId}"
+	}
+}
diff --git a/ucm2/HDA-Intel/HiFi-analog.conf b/ucm2/HDA-Intel/HiFi-analog.conf
new file mode 100644
index 0000000..1143a67
--- /dev/null
+++ b/ucm2/HDA-Intel/HiFi-analog.conf
@@ -0,0 +1,133 @@
+# Generic HDA devices for analog I/O
+
+SectionDevice."Headphones" {
+	Comment "Headphones"
+
+	If.headphone_switch {
+		Condition {
+			Type ControlExists
+			Control "name='Headphone Playback Switch'"
+		}
+		True {
+			EnableSequence [
+				cset "name='Headphone Playback Switch' on"
+			]
+			DisableSequence [
+				cset "name='Headphone Playback Switch' off"
+			]
+		}
+	}
+
+	Value {
+		PlaybackPriority 200
+		PlaybackPCM "hw:${CardId}"
+		PlaybackMixerElem "Headphone"
+		PlaybackMasterElem "Master"
+		PlaybackVolume "Headphone Playback Volume"
+		PlaybackSwitch "Headphone Playback Switch"
+		If.jack {
+			Condition {
+				Type ControlExists
+				Control "iface=CARD,name='Headphone Mic Jack'"
+			}
+			True {
+				JackControl "Headphone Mic Jack"
+			}
+			False {
+				JackControl "Headphone Jack"
+			}
+		}
+	}
+}
+
+SectionDevice."Speaker" {
+	Comment "Speaker"
+
+	If.seq {
+		Condition {
+			Type ControlExists
+			Control "name='Bass Speaker Playback Switch'"
+		}
+		True {
+			EnableSequence [
+				cset "name='Speaker Playback Switch' on"
+				cset "name='Bass Speaker Playback Switch' on"
+			]
+
+			DisableSequence [
+				cset "name='Speaker Playback Switch' off"
+				cset "name='Bass Speaker Playback Switch' off"
+			]
+		}
+		False {
+			EnableSequence [
+				cset "name='Speaker Playback Switch' on"
+			]
+
+			DisableSequence [
+				cset "name='Speaker Playback Switch' off"
+			]
+		}
+	}
+
+	Value {
+		PlaybackPriority 100
+		PlaybackPCM "hw:${CardId}"
+		PlaybackMixerElem "Speaker"
+		PlaybackMasterElem "Master"
+		PlaybackVolume "Speaker Playback Volume"
+		PlaybackSwitch "Speaker Playback Switch"
+	}
+}
+
+If.monomic {
+	Condition {
+		Type ControlExists
+		Control "name='Input Source'"
+		ControlEnum "Headphone Mic"
+	}
+	True {
+		SectionDevice."Mic2" {
+			Comment "Headphones Stereo Microphone"
+
+			ConflictingDevice [
+				"Headset"
+			]
+
+			EnableSequence [
+				cset "name='Input Source' 'Headphone Mic'"
+			]
+
+			Value {
+				CapturePriority 200
+				Include.value.File "/HDA-Intel/HDA-Capture-value.conf"
+				JackControl "Headphone Mic Jack"
+			}
+		}
+
+		SectionDevice."Headset" {
+			Comment "Headset Mono Microphone"
+
+			EnableSequence [
+				cset "name='Input Source' 'Headset Mic'"
+			]
+
+			Value {
+				CapturePriority 300
+				Include.value.File "/HDA-Intel/HDA-Capture-value.conf"
+				JackControl "Headphone Mic Jack"
+			}
+		}
+	}
+	False {
+		SectionDevice."Mic2" {
+			Comment "Headphones Stereo Microphone"
+
+			Value {
+				CapturePriority 200
+				Include.value.File "/HDA-Intel/HDA-Capture-value.conf"
+				JackControl "Mic Jack"
+			}
+		}
+	}
+}
diff --git a/ucm2/HDA-Intel/HiFi.conf b/ucm2/HDA-Intel/HiFi.conf
new file mode 100644
index 0000000..2d18c9e
--- /dev/null
+++ b/ucm2/HDA-Intel/HiFi.conf
@@ -0,0 +1,32 @@
+# Use case Configuration for sof-hda-dsp
+
+SectionVerb {
+	Value.TQ "HiFi"
+}
+
+If.analog {
+	Condition {
+		Type ControlExists
+		Control "name='Master Playback Switch'"
+	}
+	True.Include.analog.File "/HDA-Intel/HiFi-analog.conf"
+}
+
+If.hdmi {
+	Condition { Type String Empty "" }
+	True.Include.hdmi.File "/HDA-Intel/Hdmi.conf"
+}
+
+If.acp {
+	Condition {
+		Type String
+		Empty "${var:AcpCardId}"
+	}
+	True {
+		RenameDevice."Mic1" "Mic"
+	}
+	False.Include.acp {
+		Before.SectionDevice "Mic1"
+		File "/HDA-Intel/HiFi-acp.conf"
+	}
+}
diff --git a/ucm2/HDA-Intel/init.conf b/ucm2/HDA-Intel/init.conf
new file mode 100644
index 0000000..1a351a9
--- /dev/null
+++ b/ucm2/HDA-Intel/init.conf
@@ -0,0 +1,39 @@
+If.amute {
+	Condition {
+		Type ControlExists
+		Control "name='Auto-Mute Mode'"
+	}
+	True.BootSequence [
+		cset "name='Auto-Mute Mode' off"
+	]
+}
+
+If.master {
+	Condition {
+		Type ControlExists
+		Control "name='Master Playback Volume'"
+	}
+	True.BootSequence [
+		cset "name='Master Playback Volume' 60%"
+	]
+}
+
+If.speaker {
+	Condition {
+		Type ControlExists
+		Control "name='Speaker Playback Volume'"
+	}
+	True.BootSequence [
+		cset "name='Speaker Playback Volume' 60%"
+	]
+}
+
+If.headphone {
+	Condition {
+		Type ControlExists
+		Control "name='Headphone Playback Volume'"
+	}
+	True.BootSequence [
+		cset "name='Headphone Playback Volume' 60%"
+	]
+}
diff --git a/ucm2/codecs/hda/hdmi.conf b/ucm2/codecs/hda/hdmi.conf
new file mode 100644
index 0000000..c22f615
--- /dev/null
+++ b/ucm2/codecs/hda/hdmi.conf
@@ -0,0 +1,25 @@
+If.hdmi {
+	Condition {
+		Type ControlExists
+		Control "iface=CARD,name='HDMI/DP,pcm=${var:HdmiPCM} Jack'"
+	}
+	True {
+		SectionDevice."HDMI${var:HdmiNum}" {
+			Comment "HDMI${var:HdmiNum}/${var:HdmiNum} Output"
+
+			EnableSequence [
+				cset "name='IEC958 Playback Switch' on"
+			]
+
+			DisableSequence [
+				cset "name='IEC958 Playback Switch' off"
+			]
+
+			Value {
+				PlaybackPriority "${var:HdmiPrio}"
+				PlaybackPCM "hw:${CardId},${var:HdmiPCM}"
+				JackControl "HDMI/DP,pcm=${var:HdmiPCM} Jack"
+			}
+		}
+	}
+}
diff --git a/ucm2/module/lib/linked.conf b/ucm2/module/lib/linked.conf
new file mode 100644
index 0000000..d7656a8
--- /dev/null
+++ b/ucm2/module/lib/linked.conf
@@ -0,0 +1,3 @@
+ValueDefaults {
+	Linked 1
+}
diff --git a/ucm2/module/snd_acp3x_rn.conf b/ucm2/module/snd_acp3x_rn.conf
new file mode 120000
index 0000000..b56794a
--- /dev/null
+++ b/ucm2/module/snd_acp3x_rn.conf
@@ -0,0 +1 @@
+lib/linked.conf
\ No newline at end of file
-- 
2.17.1

